runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+                                  Duration pollPeriod, int pollBatchSize, int pollThreshold) {
+        Preconditions.checkArgument(pollBatchSize > 1, "pollBatchSize=%s has to be > 1",
+                pollBatchSize);
+        Preconditions.checkArgument(pollThreshold > 1, "pollThreshold=%s has to be > 1",
+                pollThreshold);
+        Preconditions.checkArgument(pollPeriod.toMillis() > 1, "pollPeriod=%s has to be > 1ms",
+                pollPeriod.toMillis());
+
+        this.scheduler = scheduler;
+        this.workers = workers;
+        this.pollPeriod = pollPeriod;
+        this.pollBatchSize = pollBatchSize;
+        this.pollThreshold = pollThreshold;
+        this.runtime = runtime;
+        this.sequencerView = runtime.getSequencerView();
+        tick();
+    }
+
+    public class Tick implements Runnable {
+        @Override
+        public void run() {
+            try {
+                schedule();
+            } catch (InterruptedException ie) {
+                log.warn("Scheduler thread has been interrupted!", ie);
+                Thread.currentThread().interrupt();
+            } catch (Throwable t) {
+                log.error("Unexpected throwable!", t);
+            }
+        }
+    }
+
+    private void tick() {
+        scheduler.submit(new Tick());
+    }
+
+    public void addTask(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                        @Nonnull String streamTag, @Nonnull List<String> tablesOfInterest,
+                        long lastAddress, int bufferSize) {
+
+        Preconditions.checkArgument(bufferSize >= pollThreshold);
+        synchronized (allTasks) {
+            if (allTasks.containsKey(streamListener)) {
+                // Multiple subscribers subscribing to same namespace and table is allowed
+                // as long as the hashcode() and equals() method of the listeners are different.
+                throw new StreamingException(
+                        "StreamingManager::subscribe: listener already registered " + streamListener);
+            }
+            StreamingTask task = new StreamingTask(runtime, workers, namespace, streamTag, streamListener,
+                    tablesOfInterest, lastAddress, bufferSize);
+            allTasks.put(streamListener, task);
+            log.debug("addTask: added {} {} {}", streamListener, namespace, streamTag);
+            allTasks.notifyAll();
+        }
+    }
+
+    public void removeTask(@Nonnull StreamListener streamListener) {
+        synchronized (allTasks) {
+            allTasks.remove(streamListener);
+            allTasks.notifyAll();
+        }
+    }
+
+    // hole filling mitigation
+    // sequencer regression mitigation
+    // fairness?
+
+    private void waitForTasks() throws InterruptedException {
+        synchronized (allTasks) {
+            while (allTasks.isEmpty()) {
+                allTasks.wait();
```

Comment:
do we really need such low-level concurrent primitives? I'm sure CompletableFutures shall be used

Reply:
What is the disadvantage of using this API ?

Reply:
It's low level, it requires a good understanding of JMM, "happens before" and other concurrency aspects. It's relatively easy to make a mistake in the code and super hard to find a possible dead-lock in the code especially if it happens in production.  I don't see why we need that level of complexity here.  
Especially when we can use a [specific library](https://github.com/ReactiveX/RxJava) for that with a clean interface, very well tested and used by major companies in the world.

Reply:
The API is low level, but the usage is straightforward and succinct. Using RxJava just to implement a [conditional variable](https://en.wikipedia.org/wiki/Monitor_(synchronization)#:~:text=A%20condition%20variable%20essentially%20is,access%20and%20resuming%20their%20task.)  seems unnecessary to me. 

I have a single scheduler thread that needs to wake up if there is at least one task that has been registered. How would that implementation look like in RxJava ?